# Overview

This set of scripts has everything required to perform association
mapping using the R package "GENESIS". GENESIS requires a genetic
relatedness matrix as a covariate. The association mapping is
performed on each of the 100 randomly drawn genotype VCF files, and
100 or 1000 permutations of the phenotypes are performed. The VCF is also annotated to perform functional enrichment analysis for diapause associated SNPs

## generate master info about each SNP

* hwe\_replace\_filters.R" calculates a number of statistics for each
SNP and uses these to generate filters for which SNPs will go into the
association mapping, and which SNPs need additional permutations

## make GDS and compute GRM for each imputation

* use "grm\_nonloco\_allsnps\_bypop.R" to calculate GRMs for populations A and B separated
* use "grm\_nonoloco\_allsnps.R to calculate GRM" for the combined A and B data

## run gwas for 100 imputations of raw data + 100 or 1000 permutations

* calculate PCAs to be used in LASSO model: make\_PCA\_for\_LASSO.R
* generate input (file names, permutation numbers, and random seeds)
with "make\_universal\_input.R"
* use "run\_universal\_mapping.sh" to run "universal\_genesis\_mapping.R" for all imputations/permutations
* the script above also runs LASSO models and calculates genomic inflation factor (GIF)

## calculate heritability

* the beginning of heritability.R makes phenotype files to be read by GCTA
* "GCTA\_heritability_perms.sh" makes a GRM in GCTA and uses that to calculate heritability in the original data and  permuted phenotype files
* the second half of heritability.R processes the heritability files generated by GCTA

## Annotate the hybrid vcf

* "snp\_annotation\_snpEff.sh" produces an annoted vcf of the parents
* "gather\_annotations.R" extracts annotations from a gds of the vcftools
* "annotation\_enrichment" performs the enrichment analysis for annotation classes (synonymous, non-syn, etc)

## miscellaneous
* "inversion\_glms.R" analyzes the effect of inversions on diapause
* "timeless\_genotypes.R" extracts genotype information about the previously published indel in timeless using parental VCF (this polymorphism was not included in our GWAS)
